### ISSUE TITLE
Adopt `typing_extensions.Override`

### ISSUE THREAD

**Issue by carmocca** [Lightning-AI/lightning#18695]
### Description & Motivation

We use overrides heavily.

Python 3.12 added support for this. See https://peps.python.org/pep-0698/

### Pitch

Use https://typing-extensions.readthedocs.io/en/latest/#override in our overrides classes.

Using this would require increasing the minimum `typing_extensions` to 4.4: https://github.com/Lightning-AI/lightning/blob/master/requirements/fabric/base.txt#L8

**Comment by seanbethard** [comment]
can you clarify what overrides classes are?

**Comment by carmocca** [comment]
@seanbethard see https://peps.python.org/pep-0698

@SarthakJain333 Sure! I suggest that you do it for 1 or a few files per PR to avoid having it blow up in size

**Comment by SarthakJain333** [comment]
Ok, I'll have a look around this framework and will try to solve the issue. Thanks for assigning me !

**Comment by JustinGoheen** [comment]
@SarthakJain333 let me know if you need any help!

**Comment by seanbethard** [comment]
I understand what overrides and decorators are. it's not clear from the issue which classes you're targeting. good luck @SarthakJain333.

**Comment by carmocca** [comment]
Sorry @seanbethard, I didn't understand what you meant from your previous message.

It would be subclasses of the following base interfaces:

- https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/accelerators/accelerator.py
- https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/loggers/logger.py
- https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/plugins/environments/cluster_environment.py
- https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/plugins/io/checkpoint_io.py
- https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/plugins/precision/precision.py
- https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/strategies/strategy.py
- https://github.com/Lightning-AI/lightning/blob/master/src/lightning/pytorch/accelerators/accelerator.py
- https://github.com/Lightning-AI/lightning/blob/master/src/lightning/pytorch/callbacks/callback.py
- https://github.com/Lightning-AI/lightning/blob/master/src/lightning/pytorch/callbacks/progress/progress_bar.py
- https://github.com/Lightning-AI/lightning/blob/master/src/lightning/pytorch/core/datamodule.py
- https://github.com/Lightning-AI/lightning/blob/master/src/lightning/pytorch/core/module.py
- https://github.com/Lightning-AI/lightning/blob/master/src/lightning/pytorch/loggers/logger.py
- https://github.com/Lightning-AI/lightning/blob/master/src/lightning/pytorch/loops/loop.py
- https://github.com/Lightning-AI/lightning/blob/master/src/lightning/pytorch/plugins/layer_sync.py
- https://github.com/Lightning-AI/lightning/blob/master/src/lightning/pytorch/plugins/precision/precision_plugin.py
- https://github.com/Lightning-AI/lightning/blob/master/src/lightning/pytorch/profilers/profiler.py
- https://github.com/Lightning-AI/lightning/blob/master/src/lightning/pytorch/strategies/strategy.py

As you can see, we make heavy use of overrides in this project

**Comment by sbshah97** [comment]
Hey if this is not done yet, is this something I can pick up? I am a new contributor looking to help.

**Comment by carmocca** [comment]
Yes @sbshah97, you can pick a file from the list above

**Comment by carmocca** [comment]
@seanbethard This is feedback for #18866 but commenting here for everybody else who decides to help.

This addition should only be done to the `src/lightning/fabric/*` and `src/lightning/pytorch/*` directories. `examples/`, `docs/`, `tests/`, or any other directories should not have this decorator added.

**Comment by sbshah97** [comment]
I can volunteer to start some files in the fabric directory to start off. Will try to avoid a large PR.

**Comment by seanbethard** [comment]
I see. well, `src/lightning/fabric/*` and `src/lightning/pytorch/*` are done. I removed overrides from `tests` following @awaelchli's feedback. happy to remove from the other directories as well.

each commit in https://github.com/Lightning-AI/lightning/pull/18866 is associated with a single base class. I thought this way of breaking up the work was preferable to adding 17 PRs.

**Comment by seanbethard** [comment]
removed overrides from all directories except `src/lightning/fabric` and `src/lightning/pytorch`. should be all set. please let me know if I missed anything.

**Comment by VictorPrins** [comment]
@seanbethard You've covered a lot of classes in #18866. I believe https://github.com/Lightning-AI/lightning/blob/master/src/lightning/pytorch/strategies/strategy.py (among others) is still unchanged. Can I do that one?

**Comment by seanbethard** [comment]
there were a few that I didn't think required any changes because the subclasses only implement abstract methods or define new ones. I believe pytorch `Strategy` is one such case.

**Comment by VictorPrins** [comment]
As an example, `setup()` is defined in `Strategy`, and overridden in `DDPStrategy` https://github.com/Lightning-AI/lightning/blob/31b8777350f545fe7b366ea7ae26d63f87e820df/src/lightning/pytorch/strategies/ddp.py#L150

Am I missing something?

**Comment by seanbethard** [comment]
nice catch! for each of the base classes listed above I only checked for classes that inherit directly from them (children). I didn't check for grandchildren or additional generations.

**Comment by VictorPrins** [comment]
Alright! I'll cover `pytorch/strategies`.

**Comment by carmocca** [comment]
@VictorPrins I might have missed some files/directories on my comment. Feel free to correct me and add the decorator elsewhere if it's needed

**Comment by awaelchli** [comment]
@VictorPrins do you have an overview of which files are missing? Or are we done?

**Comment by Borda** [comment]
> @VictorPrins do you have an overview of which files are left to do? Or are we done?

this shall be visible from `mypy` of who did we validate the changes?

**Comment by VictorPrins** [comment]
Yes I'll finish this up shortly, thanks for the heads up. The appropriate command is `mypy --enable-error-code explicit-override --follow-imports=silent <folder to check>`

**Comment by carmocca** [comment]
@VictorPrins Do you think that setting should be added to our mypy config after all the relevant files have been updated?

**Comment by VictorPrins** [comment]
Automated checking might be desirable to avoid gradual slipping, especially since the `@override` decorator is still rare enough in Python codebases that it's easy to forget. That said, there are some issues that must be navigated:

- Some folders must be excluded from the check (e.g. the `demos` folder)
- Mypy raises some errors for cases where, in my view, the `@override` decorator isn't necessary. In particular, an error is raised for overrides of methods from `builtins.object`, such as `__str__` and `__getstate__`. The error can be silenced with `# type: ignore[explicit-override]`, or solved by using the `@override` decorator for such methods. 
- Mypy raises a strange error for overrides of overloaded methods. For instance for line 89 in the code below, I get ` error: Method "precision_plugin" is not using @override but is overriding a method in class "lightning.pytorch.strategies.strategy.Strategy`, even though the `@override` is clearly used. This only happens for overloaded methods.
https://github.com/Lightning-AI/pytorch-lightning/blob/628ee0cb61637d01a44884e0d805e238ea6b49bb/src/lightning/pytorch/strategies/xla.py#L89-L91

**Comment by RecreationalMath** [comment]
Hi maintainers, quick status check on this issue:

The `@override` decorator rollout looks complete (888 usages across 120 files in `src/lightning/fabric/` and `src/lightning/pytorch/`, all 23 referenced PRs merged).

The only unresolved thread was enabling `enable_error_code = ["explicit-override"]` in mypy config to prevent regression, which stalled on @VictorPrins's concerns about false positives with `@overload` and noisy errors for `builtins.object` overrides.

Is there any pending action here, or can this be closed (optionally with a follow-up issue for the mypy enforcement)?
